test/e2e/jobscheduling.go

```diff
@@ -0,0 +1,225 @@
+
+	var kubeConfig *restclient.Config
+	var fedClient fedclientset.Interface
+	var namespace string
+	var clusterNames []string
+
+	BeforeEach(func() {
+		clusterNames = f.ClusterNames(userAgent)
+		if framework.TestContext.TestManagedFederation {
+			fixture := managed.NewSchedulingControllerFixture(tl, schedulingKind, f.ControllerConfig(), typeConfigs)
+			f.RegisterFixture(fixture)
+		} else if framework.TestContext.InMemoryControllers {
+			fixture := managed.NewSchedulingManagerFixture(tl, f.ControllerConfig())
+			f.RegisterFixture(fixture)
+		}
+		kubeConfig = f.KubeConfig()
+		fedClient = f.FedClient(userAgent)
+		namespace = f.TestNamespaceName()
+	})
+
+	testCases := map[string]struct {
+		totalParallelism    int32
+		totalCompletions    int32
+		cluster1Parallelism int32
+		cluster1Completions int32
+		cluster2Parallelism int32
+		cluster2Completions int32
+	}{
+		"Distribution happens evenly across all clusters": {
+			totalParallelism:    int32(4),
+			totalCompletions:    int32(4),
+			cluster1Parallelism: int32(2),
+			cluster1Completions: int32(2),
+			cluster2Parallelism: int32(2),
+			cluster2Completions: int32(2),
+		},
+	}
+
+	for i := range typeConfigs {
+		typeConfig := typeConfigs[i]
+
+		schedulingType := schedulingtypes.GetSchedulingType(typeConfig.GetObjectMeta().Name)
+		if schedulingType == nil || schedulingType.Kind != schedulingKind {
+			continue
+		}
+
+		templateKind := typeConfig.GetTemplate().Kind
+
+		Describe(fmt.Sprintf("scheduling for %s", templateKind), func() {
+			for testName, tc := range testCases {
+				It(fmt.Sprintf("should result in %s", testName), func() {
+					clusterCount := len(clusterNames)
+					if clusterCount != 2 {
+						framework.Skipf("Tests of JobSchedulingPreferences requires 2 clusters but got: %d", clusterCount)
+					}
+
+					var jspSpec fedschedulingv1a1.JobSchedulingPreferenceSpec
+					jspSpec = jspSpecWithoutClusterWeights(tc.totalParallelism, tc.totalCompletions)
+
+					expected := &FakeScheduleResult{}
+					expected.Result = map[string]schedulingtypes.ClusterJobValues{
+						clusterNames[0]: {
+							Completions: tc.cluster1Completions,
+							Parallelism: tc.cluster1Parallelism,
+						},
+						clusterNames[1]: {
+							Completions: tc.cluster1Completions,
+							Parallelism: tc.cluster1Parallelism,
+						},
+					}
+
+					name, err := createJSPTestObjs(fedClient, typeConfig, kubeConfig, jspSpec, namespace)
+					if err != nil {
+						tl.Fatalf("Creation of test objects failed in federation: %v", err)
+					}
+
+					err = waitForMatchingPlacement(tl, typeConfig, kubeConfig, name, namespace, expected)
+					if err != nil {
+						tl.Fatalf("Failed waiting for matching placements: %v", err)
+					}
+
+					err = waitForMatchingOverride(tl, typeConfig, kubeConfig, name, namespace, expected)
+					if err != nil {
+						tl.Fatalf("Failed waiting for matching overrides: %v", err)
+					}
+				})
+			}
+		})
+		// This test runs only for one type/kind
```

Comment:
The replica scheduling logic is already a bit funky in looping over all the type configs, and I think this break statement takes it too far.  Please separate detection of the target type configs from the test definition to avoid this hackiness.